lightning/src/util/time.rs

```diff
@@ -0,0 +1,172 @@
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! A very simple serialization framework which is used to serialize/deserialize messages as well
+//! as ChannelsManagers and ChannelMonitors.
+
+use core::ops::Sub;
+use core::time::Duration;
+use core::cell::Cell;
+/// A measurement of time.
+pub trait Time: Copy + Sub<Duration, Output = Self> where Self: Sized {
+	/// Returns an instance corresponding to the current moment.
+	fn now() -> Self;
+	
+	/// Returns the amount of time elapsed since `self` was created.
+	fn elapsed(&self) -> Duration;
+	
+	/// Returns the amount of time passed between `earlier` and `self`.
+	fn duration_since(&self, earlier: Self) -> Duration;
+	
+	/// Returns the amount of time passed since the beginning of [`Time`].
+	///
+	/// Used during (de-)serialization.
+	fn duration_since_epoch() -> Duration;
+}
+
+/// A state in which time has no meaning.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Eternity;
```

Comment:
I'd really rather not make this `pub`. One (kinda dirty) hack to avoid that while re-using the module in both crates is to symlink the file file in both crates.

Reply:
How about moving it lightning::util::test_utils?(see last commit https://github.com/lightningdevkit/rust-lightning/pull/1418/commits/99d3203cb2767bb2e8473b0f94a853fd13633a44)

Reply:
Hmm, I'd kinda still prefer we keep the trait sealed (which I *think* requires the symlink hack, but maybe there's a way to use a foreign sealed trait?).

I'm not sure that we have an immediate use-case for users implementing their own `Time`, and even if we wanted users to be able to cancel payments on no-std we'd probably want to do it via some kind of "abandon payment now" API, not with reentrancy calling user code on a `Time` trait. Additionally, we can't capture traits with associated types in bindings, so we'd have to make it non-exported there anyway.

Reply:
right. There is only one `utils.rs` instead of `utils/mod.rs` under lightning-invoice, So to avoid too much code change. I simply symlink it as `time_utils.rs`. Please let me know if it's reasonable?